### ISSUE TITLE
LibTest/io/RawDatagramSocket/timeout_A02_t01.dart should be made more robust

### ISSUE THREAD

**Issue by mkustermann** [dart-lang/co19#170]
The test sometimes fails with:
```
  FAILED: dartk-vm release_x64 co19_2/LibTest/io/RawDatagramSocket/timeout_A02_t01
  Expected: Pass
  Actual: RuntimeError
  --- Command "vm" (took 01.000002s):
  DART_CONFIGURATION=ReleaseX64 out/ReleaseX64/dart --sync-async --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/timeout_A02_t01.dart
  exit code:
  255
  stdout:
  unittest-suite-wait-for-done
  stderr:
  Unhandled exception:
  deepEquals([RawSocketEvent.write, RawSocketEvent.read, RawSocketEvent.read, RawSocketEvent.read, RawSocketEvent.closed], [RawSocketEvent.write, RawSocketEvent.read, RawSocketEvent.read, RawSocketEvent.closed], ) fails
     [cause: Expect.equals(expected: <5>, actual: <4>, 'Collections' lengths are not equal: expected length=5, actual length=4') fails.]
  #0      _fail (file:///b/s/w/ir/tests/co19_2/src/Utils/expect.dart:20:5)
  #1      Expect.deepEquals (file:///b/s/w/ir/tests/co19_2/src/Utils/expect_common.dart:375:7)
  #2      Expect.listEquals (file:///b/s/w/ir/tests/co19_2/src/Utils/expect_common.dart:303:5)
  #3      check.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/timeout_A02_t01.dart:61:16)
...
  #14     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:172:5)
```

The test should only increase the sent counter via `sent++` if the producer actually has sent something, which requires checking the return value of `producer.send()`.

Ideally the test would listen to the socket events on `producer` and only send data when getting a `write` event.

**Comment by mraleph** [comment]
Same issue with `co19_2/LibTest/io/RawDatagramSocket/timeout_A06_t01`

I think all tests in `RawDatagramSocket` category need to be audited and fixed to check the return value from the `send` method.

**Comment by mraleph** [comment]
I think RawDatagramSocket/any_A03_t01 is also suffering from the same issue.

**Comment by sgrekhov** [comment]
LibTest/io/RawDatagramSocket/timeout_A06_t01 still failing (sometimes)

**Comment by mraleph** [comment]
Similarly any_A01_t02 from the same folder has failed too. 

```dart
FAILED: dartk-vm release_x64 co19_2/LibTest/io/RawDatagramSocket/any_A01_t02
Expected: Pass
Actual: RuntimeError

--- Command "vm" (took 547ms):
DART_CONFIGURATION=ReleaseX64 out/ReleaseX64/dart --sync-async --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/any_A01_t02.dart

exit code:
255

stdout:
unittest-suite-wait-for-done

stderr:
Unhandled exception:
Expect.equals(expected: , actual: ) fails.
#0      _fail (file:///b/s/w/ir/tests/co19_2/src/Utils/expect.dart:21:5)
#1      Expect.equals (file:///b/s/w/ir/tests/co19_2/src/Utils/expect_common.dart:30:5)
#2      check... (file:///b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/any_A01_t02.dart:51:16)
#3      _RootZone.runUnary (dart:async/zone.dart:1379:54)
#4      _FutureListener.handleValue (dart:async/future_impl.dart:129:18)
#5      Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:642:45)
#6      Future._propagateToListeners (dart:async/future_impl.dart:671:32)
#7      Future._complete (dart:async/future_impl.dart:476:7)
#8      Stream.any. (dart:async/stream.dart:887:18)
#9      _RootZone.runGuarded (dart:async/zone.dart:1302:10)
#10     _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:389:13)
#11     _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:399:15)
#12     _BufferingStreamSubscription._close (dart:async/stream_impl.dart:283:7)
#13     _SyncStreamController._sendDone (dart:async/stream_controller.dart:771:19)
#14     _StreamController._closeUnchecked (dart:async/stream_controller.dart:628:7)
#15     _StreamController.close (dart:async/stream_controller.dart:621:5)
#16     new _RawDatagramSocket. (dart:io/runtime/binsocket_patch.dart:1816:23)
#17     _NativeSocket.multiplex (dart:io/runtime/binsocket_patch.dart:899:39)
#18     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)

--- Re-run this test:
python tools/test.py -m release -n dartk-linux-release-x64 co19_2/LibTest/io/RawDatagramSocket/any_A01_t02
```

I am going to mark the test as flaky again.

**Comment by dcharkes** [comment]
Similarly `distinct_A01_t05`, `asBroadcastStream_A01_t02`, and `timeout_A02_t01` from the same folder also sometimes fail.

I will mark these tests as flaky.

```dart
FAILED: dartk-vm debug_ia32 co19_2/LibTest/io/RawDatagramSocket/distinct_A01_t05
Expected: Pass
Actual: RuntimeError

--- Command "vm" (took 06.000112s):
DART_CONFIGURATION=DebugIA32 out/DebugIA32/dart --no-background-compilation --sync-async --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/distinct_A01_t05.dart

exit code:
255

stdout:
unittest-suite-wait-for-done

stderr:
Unhandled exception:
Expect.equals(expected: <3>, actual: <8>) fails.
#0      _fail (file:///b/s/w/ir/tests/co19_2/src/Utils/expect.dart:21:5)
#1      Expect.equals (file:///b/s/w/ir/tests/co19_2/src/Utils/expect_common.dart:30:5)
#2      main... (file:///b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/distinct_A01_t05.dart:61:16)
#3      _RootZone.runGuarded (dart:async/zone.dart:1302:10)
#4      _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:389:13)
#5      _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:399:15)
#6      _BufferingStreamSubscription._close (dart:async/stream_impl.dart:283:7)
#7      _ForwardingStream._handleDone (dart:async/stream_pipe.dart:106:10)
#8      _ForwardingStreamSubscription._handleDone (dart:async/stream_pipe.dart:172:13)
#9      _RootZone.runGuarded (dart:async/zone.dart:1302:10)
#10     _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:389:13)
#11     _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:399:15)
#12     _BufferingStreamSubscription._close (dart:async/stream_impl.dart:283:7)
#13     _SyncBroadcastStreamController._sendDone. (dart:async/broadcast_stream_controller.dart:401:22)
#14     _BroadcastStreamController._forEachListener (dart:async/broadcast_stream_controller.dart:328:15)
#15     _SyncBroadcastStreamController._sendDone (dart:async/broadcast_stream_controller.dart:400:7)
#16     _BroadcastStreamController.close (dart:async/broadcast_stream_controller.dart:276:5)
#17     _AsBroadcastStreamController.close (dart:async/broadcast_stream_controller.dart:504:27)
#18     _RootZone.runGuarded (dart:async/zone.dart:1302:10)
#19     _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:389:13)
#20     _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:399:15)
#21     _BufferingStreamSubscription._close (dart:async/stream_impl.dart:283:7)
#22     _SyncStreamController._sendDone (dart:async/stream_controller.dart:771:19)
#23     _StreamController._closeUnchecked (dart:async/stream_controller.dart:628:7)
#24     _StreamController.close (dart:async/stream_controller.dart:621:5)
#25     new _RawDatagramSocket. (dart:io/runtime/binsocket_patch.dart:1816:23)
#26     _NativeSocket.multiplex (dart:io/runtime/binsocket_patch.dart:899:39)
#27     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)

--- Re-run this test:
python tools/test.py -m debug -a ia32 -n dartk-linux-debug-ia32 co19_2/LibTest/io/RawDatagramSocket/distinct_A01_t05
```

```dart
FAILED: dartk-vm debug_ia32 co19_2/LibTest/io/RawDatagramSocket/timeout_A02_t01
Expected: Pass
Actual: RuntimeError

--- Command "vm" (took 08.000650s):
DART_CONFIGURATION=DebugIA32 out/DebugIA32/dart --no-background-compilation --sync-async --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/timeout_A02_t01.dart

exit code:
255

stdout:
unittest-suite-wait-for-done

stderr:
Unhandled exception:
Expect.isNull(actual: ) fails.
#0      _fail (file:///b/s/w/ir/tests/co19_2/src/Utils/expect.dart:21:5)
#1      Expect.isNull (file:///b/s/w/ir/tests/co19_2/src/Utils/expect_common.dart:57:5)
#2      check... (file:///b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/timeout_A02_t01.dart:65:16)
#3      Timer._createTimer. (dart:async/runtime/libtimer_patch.dart:21:15)
#4      _Timer._runTimers (dart:isolate/runtime/libtimer_impl.dart:382:19)
#5      _Timer._handleMessage (dart:isolate/runtime/libtimer_impl.dart:416:5)
#6      _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)

--- Re-run this test:
python tools/test.py -m debug -a ia32 -n dartk-linux-debug-ia32 co19_2/LibTest/io/RawDatagramSocket/timeout_A02_t01
```

```dart
FAILED: dartk-vm debug_ia32 co19_2/LibTest/io/RawDatagramSocket/asBroadcastStream_A01_t02
Expected: Pass
Actual: Timeout

--- Command "vm" (took 02:00.000218s):
DART_CONFIGURATION=DebugIA32 out/DebugIA32/dart --no-background-compilation --sync-async --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/asBroadcastStream_A01_t02.dart

exit code:
-15

diagnostics:
Process list including children: [17686]
Trying to capture stack trace for pid 17686
PID 17686 - process
TID 17686:
#0  0xf7764cd9 __kernel_vsyscall
#1  0xf7735d0b pthread_cond_wait@@GLIBC_2.3.2
#2  0x011b12dd dart::Monitor::WaitMicros(long long)
#3  0x011b11f7 dart::Monitor::Wait(long long)
#4  0x014137fd Dart_RunLoop
#5  0x00e5296b dart::bin::RunMainIsolate(char const*, dart::bin::CommandLineOptions*)
#6  0x00e535bf dart::bin::main(int, char**)
#7  0x00e53fa4 main
#8  0xf743eaf3 __libc_start_main
#9  0x00e52021 _start
TID 17687:
#0  0xf7764cd9 __kernel_vsyscall
#1  0xf7513ee6 epoll_wait
#2  0x00e5c3bd dart::bin::EventHandlerImplementation::Poll(unsigned int)
#3  0x00e861f4 dart::bin::ThreadStart(void*)
#4  0xf7731f72 start_thread
#5  0xf751343e __clone

eu-stack: dwfl_thread_getframes tid 17686 at 0xe52020 in /b/s/w/ir/out/DebugIA32/dart: no matching address range


stdout:
unittest-suite-wait-for-done

--- Re-run this test:
python tools/test.py -m debug -a ia32 -n dartk-linux-debug-ia32 co19_2/LibTest/io/RawDatagramSocket/asBroadcastStream_A01_t02
```

**Comment by mraleph** [comment]
Similarly `LibTest/io/RawDatagramSocket/any_A01_t03` is flaking too.

**Comment by dcharkes** [comment]
```dart
FAILED: dartk-vm product_x64 co19_2/LibTest/io/RawDatagramSocket/listen_A02_t02
Expected: Pass
Actual: RuntimeError

--- Command "vm" (took 818ms):
DART_CONFIGURATION=ProductX64 xcodebuild/ProductX64/dart --sync-async --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/listen_A02_t02.dart

exit code:
255

stdout:
unittest-suite-wait-for-done

stderr:
Unhandled exception:
deepEquals([0, 1, 2], [0, null, null], ) fails
   [cause: Expect.equals(expected: <1>, actual: <null>) fails.]
#0      _fail (file:///b/s/w/ir/tests/co19_2/src/Utils/expect.dart:21:5)
#1      Expect.deepEquals (file:///b/s/w/ir/tests/co19_2/src/Utils/expect_common.dart:375:7)
#2      Expect.listEquals (file:///b/s/w/ir/tests/co19_2/src/Utils/expect_common.dart:303:5)
#3      check.<anonymous closure>.<anonymous closure>.<anonymous closure> (file:///b/s/w/ir/tests/co19_2/src/LibTest/io/RawDatagramSocket/listen_A02_t02.dart:60:16)
#4      _RootZone.runGuarded (dart:async/zone.dart:1302:10)
#5      _BufferingStreamSubscription._sendDone.sendDone (dart:async/stream_impl.dart:389:13)
#6      _BufferingStreamSubscription._sendDone (dart:async/stream_impl.dart:399:15)
#7      _BufferingStreamSubscription._close (dart:async/stream_impl.dart:283:7)
#8      _SyncStreamController._sendDone (dart:async/stream_controller.dart:771:19)
#9      _StreamController._closeUnchecked (dart:async/stream_controller.dart:628:7)
#10     _StreamController.close (dart:async/stream_controller.dart:621:5)
#11     new _RawDatagramSocket.<anonymous closure> (dart:io/runtime/binsocket_patch.dart:1816:23)
#12     _NativeSocket.multiplex (dart:io/runtime/binsocket_patch.dart:899:39)
#13     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:171:12)

--- Re-run this test:
python tools/test.py -m product -n dartk-mac-product-x64 co19_2/LibTest/io/RawDatagramSocket/listen_A02_t02
```

This test also does not check the return value of `send()`.

**Comment by sgrekhov** [comment]
This is a part of #195. Closed. Will be fixed as part of #195
